File: p1/zero de funcao/bissecao.py
import math

####--####--####--####--####--##
#Exercise 1
# def f(x):
#   return x**2 - 3

# a = -2.48453
# b = -0.74369

####--####--####--####--####--##
#Exercise 2
#def f(x):
#   return x**2 - 4*x +2 - math.log(x)
# a = 0.26115
# b = 1.83549

####--####--####--####--####--##
#Exercise 3
# def f(x):
#   return math.exp(x) - 2*x**2 +x -1.5
# a = 1.64464
# b = 2.60854

####--####--####--####--####--##
#Exercise 4
#def f(x):
#   return x**2 - 7
# a = -3.2867
# b = -2.02933

####--####--####--####--####--##
#Exercise 5
#def f(x):
#   return x**2 - 5
# a = -2.78116
# b = -1.63452

####--####--####--####--####--##
#Exercise 6
#def f(x):
#   return x**4 -2*x**3 -3*x**2 +3*x + 2
# a = -0.96755
# b = 0.27804

####--####--####--####--####--##
#Exercise 7
# def f(x):
#   return x**3 - 7*x**2 + 14*x - 7
# a = 0.32818
# b = 4.29568
####--####--####--####--####--##

###--####--####--####--####--##
# Exercise 29

# def f(x):
   # v=32.3
   # g=9.81
   # c=24.08
   # t=7.63
   # return math.log(g)-math.log(c)-math.log(v) + math.log(x) + math.log(1-math.e**(-(c/x)*t))

# a=39.09
# b=205.18
# numIteracao = 20
# err=0
###--####--####--####--####--##
###--####--####--####--####--##
# Exercise 30

# def f(x):
   # g=9.81
   # m=70.73
   # v=26.44
   # t=7.18
   # return math.log(g)-math.log(x)-math.log(v) + math.log(m) + math.log(1-math.e**(-(x/m)*t))

# a=1
# b=100
# numIteracao = 14
# err=0
###--####--####--####--####--##
###--####--####--####--####--##
# Exercise 31

# def f(x):
   # g=9.81
   # v=9.35
   # t=4.71
   # l=3.37
   # return (math.sqrt(2*g*x)*math.tanh((math.sqrt(2*g*x)/2*l)*t)/v)-1

# a=0
# b=6
# numIteracao = 15
# err=0
###--####--####--####--####--##
###--####--####--####--####--##
# Exercise 32

# def f(x):
   # q=158.06
   # g=9.81
   # b=1.56+x
   # ac=1.56*x+x**2/2
   # return 1 - ((q**2/g*ac**3)*b)
   # return -(q**2/(g*ac**3)*b) + 1
# a=0.1
# b=8
# numIteracao = 13
# err=0
###--####--####--####--####--##
###--####--####--####--####--##
# Exercise 33

# def f(x):
   # r=7.74
   # v=480.18
   # p1=math.pi*x**2
   # p2=(3*r-x)/3
   # return ((math.pi*x**2*((3*r-x)/3))/v) - 1
# a=0
# b=15.48
# numIteracao = 11
# err=0
###--####--####--####--####--##
###--####--####--####--####--##
# Exercise 34

# def f(x):
   # r=3.5
   # ps=464.59
   # pw=1000
   # vs=(4*math.pi*r**3)/3
   # v=(math.pi*x**2/3)*(3*r-x)
   # # print("vs=={}||=={}\n".format(vs,v))
   # return ((pw*(vs-v))/(ps*vs)) - 1

# r=3.5
# a=0
# b=2*r
# numIteracao = 15
# err=0
###--####--####--####--####--##
###--####--####--####--####--##
# Exercise 35

# def f(x):
#    p0=1346395
#    v=389644
#    p=3419631
#    t=1
#    return (p0*(math.e**(x))) -p + ((v/x)*((math.e**x) -1))

# a=0.1
# b=1
# numIteracao = 13
# err=0
###--####--####--####--####--##
###--####--####--####--####--##
# Exercise 36

# def f(x):
#    l=4.56
#    r=4.77
#    v=76.54
#    r2=r**2
#    tmp=r-x
#    tmp2=(r**2)-(2*r*x)+(x**2)
#    return -v+(l*( (0.5*math.pi*r2) -
#           (r2*math.asin(tmp/r)) - (tmp*math.sqrt(r2-tmp2))))

# r=4.77
# a=0
# b=r
# numIteracao = 15
# err=0
###--####--####--####--####--##
###--####--####--####--####--##
# Exercise 37

# def f(x):

#   g=9.81
#   return -(9.81 / (2*(x**2))) * (math.sinh(x) - math.sin(x)) -1.81
# a=-4.87
# b=-0.25
# numIteracao = 11
# err=0
###--####--####--####--####--##
###--####--####--####--####--##
# Exercise 38

# def f(x):
#    aa=12
#    bb=-111.8
#    cc=171.8266
#    return aa*x**2 + bb*x + cc
# a=0
# b=4.57
# numIteracao = 12
# err=0
###--####--####--####--####--##
###--####--####--####--####--##
# PROVA - 1

# def f(x):
#   return x**3 - 7*x**2 + 14*x - 7
# a=0.21106
# b=4.39613
# numIteracao = 5
# err=0
###--####--####--####--####--##
###--####--####--####--####--##
# PROVA - 6

# def f(x):
#    r=3.66
#    ps=88.63
#    pw=1000
#    vs=(4*math.pi*r**3)/3
#    v=(math.pi*x**2/3)*(3*r-x)
#    return ((pw*(vs-v))/(ps*vs)) - 1

# r=3.66
# a=0
# b=2*r
# numIteracao = 15
# err=0
###--####--####--####--####--##
###--####--####--####--####--##
# PROVA - 7

# def f(x):
#    l=1.81
#    r=1.42
#    v=3.65
#    r2=r**2
#    tmp=r-x
#    tmp2=(r**2)-(2*r*x)+(x**2)
#    return -v+(l*( (0.5*math.pi*r2) -
#           (r2*math.asin(tmp/r)) - (tmp*math.sqrt(r2-tmp2))))

# r=1.42
# a=0
# b=r
# numIteracao = 15
# err=0

###--####--####--####--####--##
###--####--####--####--####--##
# PROVA - 5

# def f(x):
#   lbnd=1.41*(10**-10)
#   n=146775524
#   xt=n*0.25
#   p1=n+1
#   p2=1 + n*math.e**((-lbnd*n -lbnd)*x)
#   #return (n+1/(1+n*math.e**(-lbnd*(n+1)*x))) - xt
#   return (p1/p2) - xt
# a=0
# b=1818
# numIteracao = 15
# err=0

###--####--####--####--####--##
###--####--####--####--####--##
# PROVA - 8

# def f(x):
#   return 1000*((math.pi*(9.8-x))/3.0)*(pow((1.21+x*(7.07-1.21)/9.8),2)+pow(7.07,2) + (1.21+x*(7.07-1.21)/9.8)*7.07) - 343.94*((math.pi*9.8)/3.0)*(pow(1.21,2)+pow(7.07,2) + 1.21*7.07)
# a=0
# b=9.8
# numIteracao = 12
# err=0

###--####--####--####--####--##

# def f(x):
#   return 1586191*(math.e)**x + (219635/x)*(math.e**(x)-1) - 3406533

# a = 0.1
# b = 1.07
# numIteracao = 12
# err = 0

###--####--####--####--####--##

# def f(x):
#   a = 7.53 * x + (x**2)/2
#   b = 7.53 + x
#   return 1 - (((145.29)**2)/9.81*(a)**3) * b

# a = 0.49
# b = 8.21
# numIteracao = 12
# err = 0

###--####--####--####--####--##

# def f(x):
#   g = 9.81
#   q = 145.29
#   A = 7.53*x + ((x*x)/2)
#   B = 7.53 + x
#   return 1 - (((q*q)/(g*(A*A*A))))*B

# a = 0.49
# b = 8.21
# numIteracao = 12
# err = 0

###--####--####--####--####--##

def f(x):
  v = 57.55
  r = 3.17
  return math.pi*(x**2)*((3*r - x)/3) - v

a = 0
b = 6.34
numIteracao = 12
err = 0

###--####--####--####--####--##

#   ZERO DE FUNCAO
#           METODO DA BISSECAO
# Entrada: estimativa inicial (a), estimativa inicial 2 (b) funcao (f), precisão (err) e numero de iteracoes (numIteracao)

def bissecao(a,b,f,err,numIteracao):
  print("f(a) * f(b) = {0}\n" .format(f(a)*f(b)) )
  if f(a) * f(b) < 0:
    midPoint = (a+b)/2
    for x in range(numIteracao):
      if (math.fabs(f(midPoint)) > err):
        midPoint = (a+b)/2
        if f(midPoint) == 0 or (math.fabs(f(midPoint)) < err):
          print("A raiz e: {0}\n\n".format(midPoint))
        else:
          if f(a) * f(midPoint) < 0:
            b = midPoint
          else:
            a = midPoint
      else:
        print("A raiz e: {0}\n\n".format(midPoint))
        break
      print("x_{0} = {1} \t f->{2}".format(x+1, midPoint, (math.fabs(f(midPoint))) ))
  else:
    print("Nao ha raiz neste intervalo!")

if __name__ == "__main__":
  #bissecao(a,b,f,err,numIteracao)
  bissecao(inferior1, superior1, f, err, numIteracao)

File: p1/bissecao.c
#include<stdio.h>
#include<math.h>

void bisection(double (*f)(double), double a, double b, int n){
    double fa = f(a);
    double fb = f(b);
    if(fa * f(b) >= 0){
        printf("Não sei dizer que f possui raiz no intervalo raiz no intervalo [%f, %f]", a, b);
        return;
    }else{
        double m =0;
        for(int i=0; i<n; i++){
           
            m = 0.5 * (a + b);
            printf("x_%d = %.16f\n", i + 1, m);
            
            if(f(a) * f(m) < 0){
                b = m;
            }else{
                a = m;
            }
        }
    }

}

int main(int argc, char *argv[]){
    /*double f(double x){
        //double e = 2.71828182845904523530287; // euler
        return pow(x,3) -7 * pow(x,2) + 14 * x -7;
    }
    double a = 0.09722;
    double b = 4.46304;
    int n = 5;*/

    /*double f(double m){
        double e = 2.71828182845904523530287; // euler
        return ((9.81*m)/25.91)*(1 - pow(e, -(25.91/m)*8.31)) - 30.83;
    }
    double a = 36.08;
    double b = 180.1;
    int n = 20;*/

    /*double f(double c){
        double e = 2.71828182845904523530287; // euler
        return ((9.81*68.88)/c)*(1 - pow(e, -(c/68.88)*9.1)) - 20.97;
    }
    double a = 1;
    double b = 100;
    int n = 14;*/

/*    double f(double h){
        //double e = 2.71828182845904523530287; // euler
        double g = 9.81;
        return sqrt(2*g*h)*tanh(sqrt(2*g*h)/(2*5.02)*6.19) - 8.23;
    }
    
    double a = 0.19;
    double b = 17.86;
    int n = 12;
*/

/*  double f(double y){
        //double e = 2.71828182845904523530287; // euler
        double g = 9.81;
        double a = 2.93*y + pow(y,2)/2;
        double b = 2.93 + y;
        double q = 196.01;
        return 1-(pow(q, 2)/(g*pow(a, 3)))*b;
    }
    double a = 0.69;
    double b = 9.99;
    int n = 12;
*/

/* double f(double h){
        //double e = 2.71828182845904523530287; // euler
        double pi = 3.14159265358979323846;
        return (pi*pow(h, 2))*((3*7.28-h)/3) - 645.67;
    }
    double a = 0;
    double b = 14.56;
    int n = 12; */

/*     double f(double h){
        //double e = 2.71828182845904523530287; // euler
        double pi = 3.14159265358979323846;
        double r = 8.71;
        double vs = (4*pi*pow(r, 3))/3;
        double v = vs - ((737.18*vs)/1000);
        return ((pi*pow(h,2))/3) * (3*r-h) - v;
    }
    double a = 0;
    double b = 17.42;
    int n = 12; */

    /*double f(double l){
        double e = 2.71828182845904523530287; // euler
        double p0 = 1500163;
        double v = 309996;
        double p = 3075393;
        return p0*pow(e, l) + (v/l)*(pow(e, l)-1) - p;
    }
    double a = 0.1;
    double b = 1;
    int n = 15;*/

    /*double r = 1.04;
    double f(double h){
        //double e = 2.71828182845904523530287; // euler
        double pi = 3.141592653589793238462643;
        double l = 4.35;        
        double v = 6.24;

        return l*(0.5*pi*pow(r,2) - pow(r,2)*asin(h/r) - h*sqrt(pow(r,2)-pow(h,2))) - v;
    }
    //colocar r - h lá em cima
    double a = 0;
    double b = r;
    int n = 15;*/

    /*double f(double w){
        //double e = 2.71828182845904523530287; // euler
        //double pi = 3.141592653589793238462643;
        double g = 9.81;       
        double t = 1;
        double xt = 1.83;

        return -((g/(2*pow(w,2)))*(sinh(w*t) - sin(w*t))) - xt;
    }
    double a = -4.24;
    double b = -0.31;
    int n = 15;*/

    /*double f(double l){
        //double e = 2.71828182845904523530287; // euler
        //double pi = 3.141592653589793238462643;

        return (179371/1000) - ((2801/25)*l) + 12*pow(l,2);
    }
    double a = 0;
    double b = 4.96;
    int n = 15;*/

    //p1
/*     double f(double x){
        //double e = 2.71828182845904523530287; // euler
        //double pi = 3.141592653589793238462643;

        return pow(x, 3) - 7*pow(x,2) + 14*x - 7;
    }
    double a = 0.04308;
    double b = 4.30541;
    int n = 15; */

    /*double f(double t){
        double e = 2.71828182845904523530287; // euler
        //double pi = 3.141592653589793238462643; //pi
        double l = 1.41 * pow(10,-10);
        double n = 203059833;

        return (n + 1) / (1 + n*pow(e, -l*(n+1)*t)) - 25;
    }
    double a = 0;
    double b = 1337;
    int n = 15; */

    /* double r = 1.22;
    double f(double h){
        //double e = 2.71828182845904523530287; // euler
        double pi = 3.14159265358979323846;
        double vs = (4*pi*pow(r, 3))/3;
        double v = vs - ((462.37*vs)/1000);
        return ((pi*pow(h,2))/3) * (3*r-h) - v;
    }
    double a = 0;
    double b = 2*r;
    int n = 15;
 */


/*     double f(double h){
        //double e = 2.71828182845904523530287; // euler
        double pi = 3.141592653589793238462643;
        double l = 5.51;        
        double v = 49.67;
        double r = 3.29;
        return l*(0.5*pi*pow(r,2) - pow(r,2)*asin(h/r) - h*sqrt(pow(r,2)-pow(h,2))) - v;
    }
    //colocar r - h lá em cima
    double a = 0;
    double b = 3.29;
    int n = 12; */

/*     double f(double w){
        //double e = 2.71828182845904523530287; // euler
        //double pi = 3.141592653589793238462643;
        double g = 9.81;       
        double t = 1.47;
        double xt = 3.97;
        return -((g/(2*pow(w,2)))*(sinh(w*t) - sin(w*t))) - xt;
    } */

/*    double f(double x){
        return 168.6972-110.72*x+12*pow(x,2);
    } 
    double a = 0;
    double b = 4.53;
    int n = 12; */

    double f(double t){
        double e = 2.71828182846; // euler
        double n = 103481111;
        double lambda = 1.41*1 / pow(10,10);
        double x = (n * 25)/ 100;
        return ( (n+1) / (1 + n * (1/exp(lambda*(n+1)*t ))) ) - x;
    }

    double a = 0;
    double b = 2530;
    int n = 12;

    bisection(f, a, b, n);
}